### PR TITLE
add functional options for new protocol client

### PR DESCRIPTION
**Pull request by wangforthinker**
Signed-off-by: allen.wq <allen.wq@alipay.com>

### Ⅰ. Describe what this PR did
add functional options as arguments for NewProtocolClient.

### Ⅱ. Does this pull request fix one issue?

NONE.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
no need.

### CODE REVIEW

**Comment by zhouhaibing089** [comment]
May I suggest this: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis?

**Comment by wangforthinker** [comment]
> May I suggest this: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis?

Good idea!  By the way,  the argument for NewProtocolClient will be more sensible and extensive.
Now I have implemented it by functional options, could you have a review again. 
ping @zhouhaibing089  @lowzj

**Comment by wangforthinker** [comment]
I define a register function which registers functional options of which argument is map[string]string, by the way, developer could easily  create protocol client without care about instance of protocol. 
ping @zhouhaibing089 @lowzj

File: pkg/protocol/http/http.go [hunk 1]
```diff
@@ -72,54 +70,12 @@ func newDefaultTransport() *http.Transport {
...
-	return v
-}
-
 var _ protocol.Client = &Client{}
 
 // Client is an implementation of protocol.Client for http protocol.
 type Client struct {
 	client    *http.Client
-	transport http.RoundTripper
+	transport *http.Transport
```
Comment (zhouhaibing089) [inline pkg/protocol/http/http.go:78]: This seems to be a bad switch to me. Do you have a strong reason of doing this?

I can see a lot of benefits of using `http.RoundTripper`, e.g a mocking roundtripper like [go-vcr][1].

[1]: https://github.com/dnaeon/go-vcr

File: pkg/protocol/http/http.go [hunk 2]
```diff
@@ -146,46 +102,104 @@ type ClientBuilder struct {
...
+		cli.transport = transport
+		cli.client = &http.Client{
+			Transport: transport,
 		}
+
+		return nil
+	}
+}
+
+func (cb *ClientBuilder) NewProtocolClient(opts ...func(client protocol.Client) error) (protocol.Client, error) {
```
Comment (zhouhaibing089) [inline pkg/protocol/http/http.go:183]: Do we still need ClientBuilder? What're the extra value added by using a builder?

**Comment by zhouhaibing089** [comment]
I have two questions, but roughly looks good to me. :)

File: pkg/protocol/http/http.go (hunk 1 above)
Comment (wangforthinker) [reply pkg/protocol/http/http.go]: Good advice, now switch to http.RoundTripper again.

File: pkg/protocol/http/http.go (hunk 2 above)
Comment (wangforthinker) [reply pkg/protocol/http/http.go:181]: ClientBuilder is to generate client for protocol client.  For some protocol, the client needs to set some specific info to initialize client by default.

**Comment by wangforthinker** [comment]
Now, the default functional option which protocol should implement is changed from func(map[string]string) to func(map[string]interface{}) for  more general.
@zhouhaibing089 @lowzj PTAL.

**Pull request merged**